### PR TITLE
feat(eval): add eval harness and oracle baseline

### PR DESCRIPTION
**Pull request by nathanvale**
## Summary
This PR introduces the **Eval Harness + Oracle Baseline** foundation for hybrid retrieval work, and hardens it based on review feedback.

It adds:
- a reusable KPI metrics module,
- a runnable eval harness (`scripts/eval/run.ts`),
- oracle comparison utilities,
- fixture-driven benchmark inputs/baseline,
- CI-friendly eval script contract (`eval:pr`, `eval:full`, `eval:baseline:update`),
- and robust tests.

It also includes follow-up fixes for CodeRabbit/staff review issues (baseline regression wiring, stdout fallback, temp-dir cleanup, stderr propagation, oracle matching quality, and documentation accuracy).

## Why
We need a stable way to measure whether retrieval/scoring changes are improving or regressing trend quality over time.  
This PR establishes that measurement layer before deeper retrieval changes land.

### New evaluation metrics library
- Added `src/lib/eval-metrics.ts` with typed KPI helpers:
- `freshnessAtK`
- `trendRecallAtK`
- `momentumPrecisionAtK`
- `crossSourceConfirmation`
- `citationValidity`
- `runReliability`
- `performanceP95`
- `medianRunCost`
- `watchlistDeltaUtility`
- `regressionSafety`

### Eval harness runner
- Added `scripts/eval/run.ts`:
- Runs topic fixtures through CLI.
- Collects/aggregates KPI scorecard.
- Supports modes:
  - `pr` (default): mock mode, citation gate observational.
  - `full`: non-mock run, citation gate required.
  - `baseline-update`: regenerates baseline fixture.
- Writes clear scorecard fields:
  - `kpis` (raw values),
  - `checks` (raw pass/fail truth),
  - `gates` (required vs observational),
  - `passed` (gate-aware pass/fail for current mode).

### Oracle comparator
- Added `scripts/eval/oracle.ts`:
- Normalized entity matching (supports `@`, `#`, multi-word variants).
- Matches across `url/title/snippet/content` corpus (not URL-only).

### Fixtures
- Added/updated:
- `fixtures/eval/topics.json`
- `fixtures/eval/oracle.json`
- `fixtures/eval/baseline.json`
- Corrected Bun oracle entities:
  - `@bunjavascript`
  - `@jarredsumner`

### Public exports
- Updated `src/index.ts` to export eval metric types/functions.

### Script contract
- Updated `package.json`:
- `bun run eval` (backward compatible)
- `bun run eval:pr`
- `bun run eval:full`
- `bun run eval:baseline:update`

### Tests
- Added/expanded:
- `tests/eval-metrics.test.ts`
- `tests/eval-oracle.test.ts`
- Coverage includes edge cases for empty inputs, percentile math, regression threshold behavior, and normalized oracle matching semantics.

## Code Review Fixes Included
Addressed flagged issues:
- Regression check no longer compares a value to itself; it now uses baseline KPI.
- Stdout fallback runs even when outdir exists but has no JSON files.
- Temp outdirs are always cleaned in `finally`.
- Failure rows include actual stderr when available.
- Oracle matching improved beyond URL-only matching.
- Eval script aliases added and contract locked.
- Baseline snapshot semantics clarified with raw `checks` + explicit `gates`.

## Validation
Ran locally on branch:
- `bun run lint` (passes; one pre-existing warning in `src/lib/ui.ts`)
- `bun run typecheck` (pass)
- `bun run test` (pass: `166` passing)
- `bun run build` (pass)
- `bun run eval:pr` (pass)

## Known limitations (intentional for this phase)
- `eval:pr` is fixture/mock-oriented and does **not** enforce citation quality as a hard gate yet.
- `citationValidity` remains URL-format based (not full resolvability/claim verification).
- `eval:full` is intended for deeper/non-mock validation and is slower/depends on runtime environment.

## Risk
Low.
- No runtime dependency changes.
- No core retrieval/scoring pipeline behavior changes.
- Primary impact is eval harness/reporting infrastructure and correctness hardening.

### CODE REVIEW

File: scripts/eval/run.ts [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+const durations: number[] = []
+
+console.error(`Running eval for ${topics.length} topics...`)
+
+for (const { topic, category } of topics) {
+	const outdir = join(
+		'/tmp',
+		`eval-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
+	)
+	mkdirSync(outdir, { recursive: true })
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply scripts/eval/run.ts]: Implemented in `b723e03`: each temp `outdir` is now cleaned up in a `finally` block via `rmSync(..., { recursive: true, force: true })`.

File: src/lib/eval-metrics.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	k: number,
+): number {
+	if (oracle.length === 0) return 1
+	if (items.length === 0 || k <= 0) return 0
+	const sorted = [...items].sort((a, b) => b.score - a.score)
+	const topK = sorted.slice(0, k)
+	const corpus = topK.map((i) => i.url.toLowerCase()).join(' ')
+	const found = oracle.filter((entity) => corpus.includes(entity.toLowerCase()))
+	return found.length / oracle.length
+}
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply src/lib/eval-metrics.ts]: Addressed in `b723e03`: `trendRecallAtK` now normalizes entity variants (`@/#` stripping, space/hyphen/underscore/URL-encoded variants) and matches against URL + title/snippet/content corpus. Added tests in `tests/eval-metrics.test.ts`.

File: src/lib/eval-metrics.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		(c) => new Set(c.sourceTypes).size >= minSources,
+	)
+	return qualifying.length / clusters.length
+}
+
+/**
+ * Fraction of items with a valid-looking URL (starts with http:// or https://).
+ * Returns 0-1.
+ */
+export function citationValidity(items: EvalItem[]): number {
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply src/lib/eval-metrics.ts]: Aligned in `b723e03`: kept URL-focused metric semantics explicit and left KPI as format validity. Also separated required vs observational gates (`scorecard.gates`) to avoid over-trusting mock quality signals.

File: scripts/eval/run.ts [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+			results.length > 0
+				? results.reduce((a, r) => a + r.oracleRecall, 0) / results.length
+				: 0,
+	},
+	thresholds: THRESHOLDS,
+	passed: {
+		citationValidity: avgCitation >= THRESHOLDS.citationValidity,
+		runReliability: runReliability(runs) >= THRESHOLDS.runReliability,
+		performanceP95: performanceP95(durations) <= THRESHOLDS.performanceP95,
+		regressionSafety: regressionSafety(
```
[the remark this answers is not in the document]
Comment (nathanvale) [reply scripts/eval/run.ts]: Fixed in `b723e03`: regression safety now uses baseline fixture KPI (`baseline.kpis.avgCitationValidity`) instead of a self-comparison.
